components/sidebar/sidebar_channel/sidebar_channel.tsx

```diff
@@ -44,31 +45,84 @@ type Props = {
      */
     showUnreadForMsgs: boolean;
 
+    /**
+     * Gets the ref for a given channel id
+     */
+    getChannelRef: (channelId: string) => HTMLDivElement | undefined;
+
     /**
      * Sets the ref for the sidebar channel div element, so that it can be used by parent components
      */
     setChannelRef: (channelId: string, ref: HTMLDivElement) => void;
+
+    /**
+     * If category is collapsed
+     */
+    isCategoryCollapsed: boolean;
+
+    /**
+     * Is the channel the currently focused channel
+     */
+    isCurrentChannel: boolean;
 };
 
 type State = {
-
+    isUnread: boolean;
+    isCollapsed: boolean;
 };
 
 export default class SidebarChannel extends React.PureComponent<Props, State> {
-    /**
-     * Show as unread if you have unread mentions
-     * OR if you have unread messages and the channel can be marked unread by preferences
-     */
-    showChannelAsUnread = () => {
-        return this.props.unreadMentions > 0 || (this.props.unreadMsgs > 0 && this.props.showUnreadForMsgs);
-    };
+    static getDerivedStateFromProps(nextProps: Props) {
+        const isUnread = nextProps.unreadMentions > 0 || (nextProps.unreadMsgs > 0 && nextProps.showUnreadForMsgs);
```

Comment:
Instead of storing these in the derived state, could we just make a helper to calculate this from the props on demand?

Reply:
Isn't this more performant though? Since these variables are only a function of `props`, this way they are only called and updated when `props` actually changes, rather than on every `render()`

Reply:
It's actually worse, although I think the difference would end up being negligible. `getDerivedStateFromProps` is actually called whenever props are passed to the component, either normally or from `mapStateToProps`, even if nothing has changed. Here's a blog post about it, although it's talking about about `componentWillReceiveProps`: https://reactjs.org/blog/2016/01/08/A-implies-B-does-not-imply-B-implies-A.html